std/container/array.d

```diff
@@ -334,8 +342,9 @@ if (!is(Unqual!T == bool))
                 }
                 else
                 {
-                    _payload = (cast(T*) realloc(_payload.ptr,
-                            nbytes))[0 .. newLength];
+                    auto newPayloadPtr = cast(T*) realloc(_payload.ptr, nbytes);
+                    newPayloadPtr || assert(false, "std.container.Array.length failed to allocate memory.");
+                    _payload = newPayloadPtr[0 .. newLength];
```

Comment:
The other branch using `malloc` already had `newPayloadPtr || assert(false, ...)` but this branch using `realloc` didn't.

Reply:
Maybe we should trigger an `onOutOfMemoryError()` like its done in `trustedRealloc`?

Reply:
I should reverse this and go back to `assert` per https://github.com/dlang/druntime/pull/2276#discussion_r210530402

Reply:
This now uses either `assert(0)` or `onOutOfMemoryError` depending on whether version flag `D_Exceptions` is set.